airflow/cli/commands/task_command.py

```diff
@@ -280,38 +281,85 @@ def _extract_external_executor_id(args) -> str | None:
+    so that logs are still emitted to stdout.
+    """
+    # if there are no task handlers, then we should not do anything
+    # because either the handlers were already moved by the LocalTaskJob
+    # invocation of task_run (which wraps the --raw invocation), or
+    # user is doing something custom / unexpected
+    if not ti.log.handlers or settings.DONOT_MODIFY_HANDLERS:
+        yield
+        return
+    is_k8s_executor_pod = os.environ.get("AIRFLOW_IS_K8S_EXECUTOR_POD")
+    root_logger = logging.getLogger()
+    root_level = root_logger.level
+    task_logger = ti.log
+    task_propagate = task_logger.propagate
+    task_handlers = task_logger.handlers.copy()
+
+    # these are already copied to root logger, so remove and propagate
+    # (this ensures they are not handled more than they need to be)
+    # we'll add them back later
+    for h in task_logger.handlers[:]:
+        task_logger.removeHandler(h)
+    task_logger.propagate = True
+    root_logger.setLevel(task_logger.level)
+    root_handlers = root_logger.handlers.copy()
+    root_logger.handlers[:] = task_handlers
+
+    # task log handler reads from k8s pod logs when pod still running
+    # so we need to keep the console handler and set it to not respect
+    # redirection because we read live pod logging from container stdout
+    if is_k8s_executor_pod:
+        for h in root_handlers:
+            if isinstance(h, RedirectStdHandler):
+                root_logger.addHandler(h)
+                # "redirect stdout" context managers work by changing what lives at
+                # sys.stdout. If respect_redirection is True, then we reference whatever
+                # "sys.stdout" is at runtime. If False, this handler will store the value
+                # of sys.stdout at handler *init* time, and stream to *that* instead of whatever
+                # the current value is.
+                h.respect_redirection = False
+                break
+    try:
+        yield
+    finally:
+        task_logger.propagate = task_propagate
+        root_logger.setLevel(root_level)
+        root_logger.handlers[:] = root_handlers
+        task_logger.handlers[:] = task_handlers
+        if is_k8s_executor_pod:
+            for h in root_handlers:
+                if isinstance(h, RedirectStdHandler):
+                    h.respect_redirection = True
 
-    - Redirect stdout and stderr to the task instance log, as INFO and WARNING
-      level messages, respectively.
 
+@contextmanager
+def _redirect_stdout_to_ti_log(ti: TaskInstance) -> Generator[None, None, None]:
     """
-    modify = not settings.DONOT_MODIFY_HANDLERS
-    if modify:
-        root_logger, task_logger = logging.getLogger(), logging.getLogger("airflow.task")
+    Redirect stdout to ti logger.
 
-        orig_level = root_logger.level
-        root_logger.setLevel(task_logger.level)
-        orig_handlers = root_logger.handlers.copy()
-        root_logger.handlers[:] = task_logger.handlers
```

Comment:
@malthe observe, we already copy (temporarily) to root logger the thandlers from airflow.task

This causes problems because we have to have complicated propagation rules at the airflow.task logger because we leave it there.

While my solution here is a few more lines (and a lot more comments) it's not very complicated.  There are two parts.  

1. Instead of _copying_ the handlers to root, I _move_ them to root.  We don't need them at task if they are already at root.  This could ultimately allow for simplification of our propagation logic.
2. Previously we _remove_ our console handler from root at run time.  Now, if we're in a k8s executor pod, I keep the console handler there.

Reply:
I didn't quite understand what "stdout redirection" means – is that within Python or is that something happening outside of Python (in a shell script perhaps) – ?

When I think about this problem, it seems that there is an orthogonal concern in the logging setup of Airflow which is whether or not the task logger (during execution) should be emitted to the stdout stream.

That's something you could want in any situation, not just K8S.

Reply:
> I didn't quite understand what "stdout redirection" means – is that within Python or is that something happening outside of Python (in a shell script perhaps) – ?

i am not sure what comment you refer to but, what this means is there are stdlib helpers to take things that would ordinarily go to stdout (e.g. print) and send them to some other stream. 

see `from contextlib import redirect_stdout`.  it just monkey patches `sys.stdout` temporarily.

we use them (for better or worse) to redirect stdout to task logs.

our log read logic when task is running is usually, read from flask log server.  most task logging is redirected to log (and therefore file) and our "console" handler respects this redirection.   for obvious reasons -- the celery worker log, or local executor log would get unreasonably chatty.  but in k8s executor context, there is no log server on the worker, and there is no problem with keeping the stdout, and importantly our log read logic _assumes_ everything will be forwarded to stdout -- and this is what's broken that i am here fixing.